### PR TITLE
feat(CommunityPermissions): Dirty state handling improvements, API and some components simplified

### PR DESCRIPTION
**Pull request by micieslak**
### What does the PR do

- save/reset handlers refactored in `CommunityNewPErmissionView`
- dirty state handling simplified in `CommunityNewPermissionView`
- `ModelsComparator` component introduced for tracking models equality in declarative and reusable way
- some components excluded to separate files to reduce sizes of single files, inline components used
- on the fly detection of duplicated permissions implemented
- checking if permission form is fully filled
- permission model simplified - now it doesn't contain metadata of holdings like image or short name, just refers to tokens by key. UI is responsible for proper composition of information from permission models and tokens/collectibles models
- some warnings fixed

Note: This PR is big but it should be easy to go commit by commit and review it in a smaller chunks :)

Closes: #9388
Closes: #9048
Closes: #8830

### Affected areas
Components related to Community Permissions functionality
### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

[Kazam_screencast_00116.webm](https://user-images.githubusercontent.com/20650004/219098255-b5f12f5d-194a-433e-bc1b-8601cbbd3b8b.webm)

[Kazam_screencast_00117.webm](https://user-images.githubusercontent.com/20650004/219098263-c02eb4f4-6d0c-4645-931d-515fb852c996.webm)

[Kazam_screencast_00118.webm](https://user-images.githubusercontent.com/20650004/219098269-a034e301-65ec-40cd-9d58-d6dd00b99f67.webm)

[Kazam_screencast_00119.webm](https://user-images.githubusercontent.com/20650004/219098272-e68d9626-46ca-4d6b-9760-0b3dae5c03b1.webm)

### CODE REVIEW

**Comment by caybro** [comment]
@micieslak 
```
qrc:/StatusQ/Core/StatusScrollView.qml: Type ModelUtils unavailable, qrc:/StatusQ/Core/Utils/ModelUtils.qml: No such file or directory
```

**Comment by micieslak** [comment]
> @micieslak
> 
> ```
> qrc:/StatusQ/Core/StatusScrollView.qml: Type ModelUtils unavailable, qrc:/StatusQ/Core/Utils/ModelUtils.qml: No such file or directory
> ```

whoops...

**Comment by micieslak** [comment]
> @micieslak
> 
> ```
> qrc:/StatusQ/Core/StatusScrollView.qml: Type ModelUtils unavailable, qrc:/StatusQ/Core/Utils/ModelUtils.qml: No such file or directory
> ```

Fixed, thanks!

File: ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml [hunk 1]
```diff
@@ -71,7 +74,7 @@ StatusScrollView {
     // roles: itemId, text, emoji, color
     property var channelsModel: ListModel {}
 
-    property bool isPrivate
+    property bool isPrivate: false
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml]: `false` by default

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            Layout.fillWidth: true
+            StatusBaseText {
+                text: qsTr("Hide permission")
+                color: Theme.palette.directColor1
+                font.pixelSize: 15
+            }
+            StatusBaseText {
+                Layout.fillWidth: true
+                Layout.fillHeight: true
+                text: qsTr("Make this permission hidden from members who don’t meet it’s requirements")
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:35]: "its" (typo)

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import QtQuick 2.14
+import QtQuick.Controls 2.14
+import QtQuick.Layouts 1.14
+
+import StatusQ.Core 0.1
+import StatusQ.Components 0.1
+import StatusQ.Controls 0.1
+import StatusQ.Core.Theme 0.1
+
+//import utils 1.0
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:10]: commented out code?

File: ui/app/AppLayouts/Chat/views/communities/CommunityPermissionsView.qml [hunk 4]
```diff
@@ -92,27 +87,26 @@ StatusScrollView {
...
-            confirmationText: qsTr("If you delete this permission, any of your community members who rely on this permission will loose the access this permission gives them.")
-            onConfirmButtonClicked: {
-                store.removePermission(d.permissionIndexToRemove)
-                close()
-            }
+    ConfirmationDialog {
+        id: declineAllDialog
+
+        header.title: qsTr("Sure you want to delete permission")
+        confirmationText: qsTr("If you delete this permission, any of your community members who rely on this permission will loose the access this permission gives them.")
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/views/communities/CommunityPermissionsView.qml:100]: s/loose/lose

File: ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+                return CommunityPermissionsHelpers.getTokenNameByKey(model, key)
+            }
+
+            function getText(type, key, amount) {
+                const name = getName(type, key)
+
+                return CommunityPermissionsHelpers.setHoldingsTextFormat(
+                            type, name, amount)
+            }
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml:35]: imho these inner functions shouldn't be inside the `ExpressionRole`, better put them in a private `QtObject`; the _whole_ block gets evaluated by SortFilterProxyModel

File: ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            name: "imageSource"
+
+            function getIcon(type, key) {
+                if (type === HoldingTypes.Type.Ens)
+                    return Style.svg("profile/ensUsernames")
+
+                const model = type === HoldingTypes.Type.Asset
+                            ? assetsModel : collectiblesModel
+
+                return CommunityPermissionsHelpers.getTokenIconByKey(model, key)
```
Comment (caybro) [inline ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml:51]: Same here

**Review by caybro — COMMENTED** [review]
Awesome stuff, really some minor things inline

File: ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml (hunk 1 above)
Comment (micieslak) [reply ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml]: Yeah, I know, just for verbosity/readability...

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml (hunk 2 above)
Comment (micieslak) [reply ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:35]: Good point, it's also in design (I'm copying it, not rewriting ofc :D), so I fixed here left a comment in figma

File: ui/app/AppLayouts/Chat/views/communities/CommunityPermissionsView.qml (hunk 4 above)
Comment (micieslak) [reply ui/app/AppLayouts/Chat/views/communities/CommunityPermissionsView.qml:100]: same as above, fixed and commented in figma

File: ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml (hunk 5 above)
Comment (micieslak) [reply ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml:35]: Are you sure that the stuff outside `expression` property in `ExpressionRole` is treated is some special way? I would guess that's `ExpressionRole` is just an object derived from `QObject` and function defined here is not treated in any special way.

My rationale for putting it here is related to the fact that it's workaround, enforced by limitations/bugs of `expression` property. It's generally better imo when code for workaround is localized possibly close to the source of the problem.

Comment (micieslak) [reply ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml:35]: https://github.com/oKcerG/SortFilterProxyModel/blob/master/proxyroles/expressionrole.h
https://github.com/oKcerG/SortFilterProxyModel/blob/master/proxyroles/expressionrole.cpp

I haven't analysed it thoroughly but for me only expression is reevaluated by the model so there is no penalty for defining methods inside.

Comment (caybro) [reply ui/app/AppLayouts/Chat/views/communities/HoldingsSelectionModel.qml:35]: Yeah, looks like you're correct; makes sense then

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                text: qsTr("Hide permission")
+                color: Theme.palette.directColor1
+                font.pixelSize: 15
+            }
+            StatusBaseText {
+                Layout.fillWidth: true
+                Layout.fillHeight: true
+                text: qsTr("Make this permission hidden from members who don’t meet it’s requirements")
+                color: Theme.palette.baseColor1
+                font.pixelSize: 15
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:35]: Style.current.primaryTextFontSize might be better. Same with `Hide permissions`

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+Control {
+    id: root
+
+    property alias checked: switchItem.checked
+
+    signal toggled
+
+    contentItem: RowLayout {
+        spacing: 16
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:18]: spacing: Style.current.padding

File: ui/app/AppLayouts/Chat/panels/communities/PermissionDuplicationWarningPanel.qml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+
+import utils 1.0
+
+Control {
+    id: root
+
+    spacing: 8
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/panels/communities/PermissionDuplicationWarningPanel.qml:13]: spacing: Style.current.halfPadding?

File: ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml [hunk 10]
```diff
@@ -659,69 +494,51 @@ StatusScrollView {
...
             Layout.fillWidth: true
-            onClicked: {
-                root.store.createPermission(d.dirtyValues.holdingsModel,
-                                            d.dirtyValues.permissionObject,
-                                            d.dirtyValues.isPrivateDirty ? !root.isPrivate : root.isPrivate,
-                                            d.dirtyValues.channelsModel)
-                root.permissionCreated()
-            }
+            Layout.topMargin: conflictPanel.visible
+                              ? conflictPanel.Layout.topMargin : 24 // by design
```
Comment (alexjba) [inline ui/app/AppLayouts/Chat/views/communities/CommunityNewPermissionView.qml:535]: Layout.topMargin: conflictPanel.visible ? conflictPanel.Layout.topMargin : `Style.current.bigPadding`?

**Review by alexjba — COMMENTED** [review]
Really nice work! 👍 I've done some testing and LGTM
Minor inline comments on hardcoded paddings and text size.

**Review by noeliaSD — CHANGES_REQUESTED** [review]
Overall, very good job!!! Nothing to add in terms of code.

Testing it I found the following:

1. When you select an asset, the shortName should be dispalyed. Now it is displayed the name instead. 
So the general rule should be: if there is shortName, display shortName, if not, display the name.

2. We've lost the automatic switch to `on` for `hide permission` when moderate type is selected.

File: ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            StatusBaseText {
+                Layout.fillWidth: true
+                Layout.fillHeight: true
+                text: qsTr("Make this permission hidden from members who don’t meet it’s requirements")
+                color: Theme.palette.baseColor1
+                font.pixelSize: 15
+                lineHeight: 1.2
+                wrapMode: Text.WordWrap
+                elide: Text.ElideRight
+                clip: true
```
[the remark this answers is not in the document]
Comment (micieslak) [reply ui/app/AppLayouts/Chat/panels/communities/HidePermissionPanel.qml:39]: removed

File: ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml [hunk 12]
```diff
@@ -62,33 +65,40 @@ Control {
         id: column
         spacing: root.spacing
 
-        HoldingsListPanel {
-            Layout.fillWidth: true
+        component CustomHoldingsListPanel: HoldingsListPanel {
```
[the remark this answers is not in the document]
Comment (micieslak) [reply ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:69]: right, done

**Comment by micieslak** [comment]
> Overall, very good job!!! Nothing to add in terms of code.
> 
> Testing it I found the following:
> 
> 1. When you select an asset, the shortName should be dispalyed. Now it is displayed the name instead.
>    So the general rule should be: if there is shortName, display shortName, if not, display the name.
> 2. We've lost the automatic switch to `on` for `hide permission` when moderate type is selected.

Thank you @noeliaSD for review and testing! I fixed both issues.

**Review by alexjba — APPROVED** [review]

**Comment by noeliaSD** [comment]
> > Overall, very good job!!! Nothing to add in terms of code.
> > Testing it I found the following:
> > 
> > 1. When you select an asset, the shortName should be dispalyed. Now it is displayed the name instead.
> >    So the general rule should be: if there is shortName, display shortName, if not, display the name.
> > 2. We've lost the automatic switch to `on` for `hide permission` when moderate type is selected.
> 
> Thank you @noeliaSD for review and testing! I fixed both issues.

Thanks for the updates! Tested and they look fine!

During the test I've found another curious issue when long numbers are introduced (also tested on master and I cannot reproduce it there):
It also happens with other combination of numbers but the one recorded works to always reproduce it (you need to do it the first time entering a new permission):

https://user-images.githubusercontent.com/97019400/219602936-f5c8281c-4a87-40b2-848a-20cb216ea19e.mov

Here the traces: 
```
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
Warning: DelegateModel::item: index out range 4 4 (:0, )
Warning: DelegateModel::item: index out range 5 4 (:0, )
Warning: DelegateModel::item: index out range 6 4 (:0, )
````

**Comment by micieslak** [comment]
> > > Overall, very good job!!! Nothing to add in terms of code.
> > > Testing it I found the following:
> > > 
> > > 1. When you select an asset, the shortName should be dispalyed. Now it is displayed the name instead.
> > >    So the general rule should be: if there is shortName, display shortName, if not, display the name.
> > > 2. We've lost the automatic switch to `on` for `hide permission` when moderate type is selected.
> > 
> > 
> > Thank you @noeliaSD for review and testing! I fixed both issues.
> 
> Thanks for the updates! Tested and they look fine!
> 
> During the test I've found another curious issue when long numbers are introduced (also tested on master and I cannot reproduce it there): It also happens with other combination of numbers but the one recorded works to always reproduce it (you need to do it the first time entering a new permission):
> 
> Here the traces:
> 
> ```
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:209, )
> Warning: qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213:25: Unable to assign [undefined] to QString (qrc:/StatusQ/src/StatusQ/Components/StatusItemSelector.qml:213, )
> Warning: DelegateModel::item: index out range 4 4 (:0, )
> Warning: DelegateModel::item: index out range 5 4 (:0, )
> Warning: DelegateModel::item: index out range 6 4 (:0, )
> ```

Thank you @noeliaSD for point that. I also noticed that. After several attempts I gave up - I will exclude it to a separate task (in order to do other stuff first). It's a problem with `SortFilterProxyModel`.

**Review by caybro — APPROVED** [review]
Code LGTM

**Review by noeliaSD — APPROVED** [review]

**Pull request merged**